arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/SuspendConnection.kt

```diff
@@ -26,12 +26,42 @@ inline class CancelToken(val cancel: suspend () -> Unit) {
 
 typealias Disposable = () -> Unit
 
+/**
+ * Grab the [SuspendConnection] from [CoroutineContext],
+ * when none is found then default to an uncancellable connection.
+ *
+ * That could only happen when we call [connection] inside a [suspend] function **not** started by our [Environment],
+ * or [startCoroutineCancellable] runners.
+ * When Arrow Fx Coroutines code is started by KotlinX runners, then the Arrow Fx Coroutines KotlinX Coroutines module must install a [SuspendConnection]
+ * for the operation to be cancellable, and thus we'd find a connection here as well.
+ */
 internal fun CoroutineContext.connection(): SuspendConnection =
   this[SuspendConnection] ?: SuspendConnection.uncancellable
 
 /**
- * SuspendConnection is a state-machine inside [CoroutineContext] that manages cancellation.
- * This could in the future also serve as a mechanism to collect debug information on running connections.
+ * [SuspendConnection] is a [CoroutineContext] and has two implementations:
+ *   - [Uncancellable]
+ *   - [DefaultConnection]
+ *
+ * # Uncancellable
+ *
+ * The [Uncancellable] implementation has a no-op [cancel] method, and always returns [false] upon checking [isCancelled].
+ * Which subsequently means that all cancellable operations become uncancellable when run on this [CoroutineContext],
+ * and inserting [cancelBoundary] becomes a no-op operation.
+ * Some optimisations could be implemented when an [Uncancellable] context is detected,
+ * such as [bracket] could run as a simple `try/catch/finally` since [ExitCase.Cancelled] can never occur on a [Uncancellable] connection.
+ * See the [arrow.fx.coroutines.uncancellable] combinator implementation to see how you can swap from any kind of [SuspendConnection] to a [Uncancellable] and back to the original connection.
+ *
+ * # DefaultConnection
+ *
+ * [DefaultConnection] keeps track of all [CancelToken] registered for a given [CoroutineContext].
+ * It does so by keeping all [CancelToken] in a FIFO stack, and running them in order when cancelled.
+ *
+ * Since [SuspendConnection] is a FIFO stack, we can register a token before running a cancellable operation,
+ * and pop the token after the cancellable operation has finished so
+ * we don't run a [CancelToken] when it's not necessary anymore since that could lead to (undefined) weird behavior.
+ * You can see this usage in [parMapN], [raceN], [racePair] & [raceTriple].
```

Comment:
This seems inaccurate, it may lead to some open questions like:
- Why does it being a FIFO stack mean we can register before and pop after a cancellable op? This doesn't seem related.
- When is a running CancelToken no longer necessary?

Reply:
> Why does it being a FIFO stack mean we can register before and pop after a cancellable op? This doesn't seem related.

Hmm, you're correct. It's more related to how it's supposed to be used I guess. Not sure how we can phrase it more clearly.

> When is a running CancelToken no longer necessary?

For example in `parMap2`, we launch 2 tasks in parallel on an existing `SuspendConnection`.
So we create two new `SuspendConnection` for those 2 parallel tasks, and push them as a single entry on the stack of the original context.

```
val connA = SuspendConnection()
val connB = SuspendConnection()
originalContext.pushPair(connA, connB)
```

There are two scenarios where we'd want to `pop` the entry, since running the `CancelToken` is no longer necessary. When either task failed, and after we canceled the other one OR when both have completed successfully. Since in both those scenarios the tasks have already been canceled, or completed.

```
fun complete(a: A, b: B) {
  conn.pop()
  cb(
    try {
      Result.success(f(a, b))
    } catch (e: Throwable) {
      Result.failure<C>(e.nonFatalOrThrow())
    }
  )
}
fun sendException(other: SuspendConnection, e: Throwable) = when (state.getAndSet(e)) {
  is Throwable -> Unit // Do nothing we already finished
  else -> other.cancelToken().cancel.startCoroutine(Continuation(EmptyCoroutineContext) { r ->
    conn.pop()
    cb(Result.failure(r.fold({ e }, { e2 -> Platform.composeErrors(e, e2) })))
  })
}
```

Reply:
@1Jajen1 any thoughts on how we an improve?

Reply:
Not really, I think this explanation on when `CancelToken`'s are no longer necessary is good to have, but for the stack, I have no idea how to phrase it better :see_no_evil:

Reply:
No problem, I'll give this a shot to re-phrase better with the added example.